flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala

```diff
@@ -119,6 +150,57 @@ class DataStreamOverAggregate(
 
   }
 
+  def createTimeBoundedProcessingTimeOverWindow(inputDS: DataStream[Row]): DataStream[Row] = {
+
+    val overWindow: Group = logicWindow.groups.get(0)
+    val partitionKeys: Array[Int] = overWindow.keys.toArray
+    val namedAggregates: Seq[CalcitePair[AggregateCall, String]] = generateNamedAggregates
+
+    val index = overWindow.lowerBound.getOffset.asInstanceOf[RexInputRef].getIndex
+    val count = input.getRowType().getFieldCount()
+    val lowerboundIndex = index - count
+    
+    
+    val time_boundary = logicWindow.constants.get(lowerboundIndex).getValue2 match {
+      case _: java.math.BigDecimal => logicWindow.constants.get(lowerboundIndex)
+         .getValue2.asInstanceOf[java.math.BigDecimal].longValue()
+      case _ => throw new TableException("OVER Window boundaries must be numeric")
+    }
+
+     // get the output types
+    val rowTypeInfo = FlinkTypeFactory.toInternalRowTypeInfo(getRowType).asInstanceOf[RowTypeInfo]
+         
+    val result: DataStream[Row] =
+        // partitioned aggregation
+        if (partitionKeys.nonEmpty) {
+          
+          val processFunction = AggregateUtil.CreateTimeBoundedProcessingOverProcessFunction(
+            namedAggregates,
+            inputType,
+            time_boundary)
+          
+          inputDS
+          .keyBy(partitionKeys: _*)
+          .process(processFunction)
+          .returns(rowTypeInfo)
+          .name(aggOpName)
+          .asInstanceOf[DataStream[Row]]
+        } else { // non-partitioned aggregation
+          val processFunction = AggregateUtil.CreateTimeBoundedProcessingOverProcessFunction(
```

Comment:
Hi @rtudoran, IMO `MapState` is the better option. Have a look at this [comment](https://github.com/apache/flink/pull/3574#issuecomment-288646109) where I explain the benefits of the `MapState` approach.

It is true, that we need to read all keys if we use `MapState`, but 1) this is only read/deserialization 2) it is cheap `Long` values. The huge advantage of `MapState` is that we only have to reading and writing relevant `Row` values.

With `ListState` and `ValueState` we always have to read and write all `Row` values.

The complexity of operating on the deserialized structures should be very similar for all approaches and be negligible compared to the cost of de/serializing (which includes object instantiations)